csrc/kernel_cache.cpp

```diff
@@ -1584,37 +1585,102 @@ void FusionKernelRuntime::updateHeuristicsLaunchParams(
-    if (!maybe_scheduler_entry.has_value()) {
-      return std::nullopt;
+
+  // The runtime group run order is different from the segmented_fusion group
+  // order. Instead of using FusionHeuristics::emplaceBack, we initialize
+  // FusionHeuristics with the desired number of groups.
+  std::optional<FusionKernelRuntime::HeuristicsPtr> opt_heuristics;
+  const int64_t num_groups = (int64_t)runtime_workspace_.group_run_order.size();
+  opt_heuristics = std::make_unique<FusionHeuristics>(num_groups);
+
+  // Store metadata copy of arguments for ArgumentManager
+  KernelArgumentHolder args_metadata;
+  std::transform(
+      args.cbegin(),
+      args.cend(),
+      args_metadata.getBackInserter(),
+      convertMetadataArg);
+  args_metadata.setDeviceIndex(args.getDeviceIndex());
+
+  // ArgumentManager manipulates the KernelArgumentHolder argument.
+  // We make another metadata copy for the PrecomputedValues.
+  KernelArgumentHolder complete_fusion_metadata_args(args_metadata);
+
+  ArgumentManager args_manager(
+      args_metadata, runtime_workspace_, segmented_fusion_->inputs());
+
+  // Follow group run order
+  for (int64_t group_id : c10::irange(num_groups)) {
+    auto group_to_run = runtime_workspace_.group_run_order.at(group_id);
+
+    // Create fusion for this segmented group
+    Fusion* fusion_to_run = group_to_run->getFusion();
+    NVF_ERROR(fusion_to_run != nullptr);
+    FusionGuard fg(fusion_to_run);
+
+    // Get input arguments for SchedulerRuntimeInfo
+    KernelArgumentHolder group_runtime_inputs;
+    for (auto input : group_to_run->inputs()) {
+      group_runtime_inputs.push(*args_manager.checkTensorMap(input));
     }
-    auto scheduler_entry = std::move(maybe_scheduler_entry.value());
-    if (!scheduler_entry->sameAs(
-            heuristics_->heuristicsList()[group_index].get())) {
-      return std::nullopt;
+
+    // Create PrecomputedValues initialized with original fusion inputs
+    auto evaluator_precomputed_values =
+        std::make_unique<PrecomputedValues>(fusion_to_run);
+    evaluator_precomputed_values->bindInputs(group_runtime_inputs);
+    evaluator_precomputed_values->bindValues(
+        group_to_run->getCompleteFusionInputs(), complete_fusion_metadata_args);
+    evaluator_precomputed_values->evaluate();
+
+    // Get all tensorviews for segmented fusion
+    std::vector<TensorView*> all_tvs_for_fusion_to_run =
+        ir_utils::allTvs(fusion_to_run);
+
+    SchedulerRuntimeInfo fusion_to_run_info(
+        fusion_to_run,
+        group_runtime_inputs,
+        evaluator_precomputed_values.get(),
+        all_tvs_for_fusion_to_run,
+        forced_index_type);
+
+    if (initial_heuristics) {
+      // Add new scheduler entry for this segmented group
+      opt_heuristics.value()->at(group_to_run->groupId()) =
+          segmented_fusion_->makeInitialSchedulerEntry(
+              fusion_to_run, group_to_run, fusion_to_run_info);
```

Comment:
Can `makeInitialSchedulerEntry` and `getMaybeSchedulerEntry` below access `fusion_to_run` via `fusion_to_run_info->fusion()`? This way, we don't need to change their interface to take a Fusion.

Reply:
I also moved `getMaybeSchedulerEntry` back to `SegmentedGroup`.